File: project2/Dotproduct.java
import java.util.Scanner;

public class Dotproduct {
    static long pcg_state = 0x853c49e6748fea9bL;
    static long pcg_inc = 0xda3e39cb94b95bdbL;

    // Generate 32-bit unsigned random number as int
    static int pcg32_random_r() {
        long oldstate = pcg_state;
        // 64-bit LCG
        pcg_state = oldstate * -4348849565147123411L + (pcg_inc | 1L);
        // Output permutation (XSH-RR)
        int xorshifted = (int) (((oldstate >>> 18) ^ oldstate) >>> 27);
        int rot = (int) (oldstate >>> 59);
        return (xorshifted >>> rot) | (xorshifted << ((-rot) & 31));
    }

    // Wrappers for different data types
    static int generate_int() {
        long u_pcg = Integer.toUnsignedLong(pcg32_random_r());
        return (int)(u_pcg % 21) - 10;
    }

    static float generate_float() {
        long u_pcg = Integer.toUnsignedLong(pcg32_random_r());
        return (float)generate_int() + (float)(u_pcg % 1000) / 1000.0f;
    }

    static double generate_double() {
        long u_pcg = Integer.toUnsignedLong(pcg32_random_r());
        return (double)generate_int() + (double)(u_pcg % 1000000) / 1000000.0;
    }

    static byte generate_signed_char() {
        return (byte)generate_int();
    }

    static short generate_short() {
        return (short)generate_int();
    }
    
    public static void main(String[] args) {
        int[] sizes = {10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000};
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("\n==========================================");
            System.out.println("Please select the data type for calculation:");
            System.out.println("0. Exit program");
            System.out.println("1. int");
            System.out.println("2. float");
            System.out.println("3. double");
            System.out.println("4. signed char (byte in Java)");
            System.out.println("5. short");
            System.out.println("6. all");
            System.out.print("> ");

            String line = scanner.nextLine().trim();
            int type;

            try {
                type = Integer.parseInt(line);
            } catch (NumberFormatException e) {
                System.out.println("Invalid input. Please enter an integer from 0 to 6.");
                continue;
            }

            if (type == 0) {
                System.out.println("Exiting program...");
                break;
            } else if (type >= 1 && type <= 6) {
                System.out.printf("\n[ Type %d selected, running all sizes ]\n", type);

                for (int size : sizes) {
                    System.out.printf("\n--- Size: %d ---\n", size);
                    try {
                        switch (type) {
                            case 1: int_dot_product(size); break;
                            case 2: float_dot_product(size); break;
                            case 3: double_dot_product(size); break;
                            case 4: signed_char_dot_product(size); break;
                            case 5: short_dot_product(size); break;
                            case 6:
                                int_dot_product(size);
                                System.gc();
                                float_dot_product(size);
                                System.gc();
                                double_dot_product(size);
                                System.gc();
                                signed_char_dot_product(size);
                                System.gc();
                                short_dot_product(size);
                                System.gc();
                                break;
                        }
                    } catch (OutOfMemoryError e) {
                        System.out.println("Memory allocation failed for size " + size);
                        System.gc(); // Ask JVM to clean up memory
                    }
                }

                System.out.printf("\n[ All sizes computed for type %d. Waiting for next instruction. ]\n", type);
            } else {
                System.out.println("Invalid input. Please enter an integer from 0 to 6.");
            }
        }
        scanner.close();
    }
    
    static void int_dot_product(int size) {
        int[] a = new int[size];
        int[] b = new int[size];

        pcg_state = System.currentTimeMillis(); // Reseed

        for (int i = 0; i < size; i++) {
            a[i] = generate_int();
            b[i] = generate_int();
        }

        long start = System.nanoTime();
        long sum = 0;
        for (int i = 0; i < size; i++) {
            sum += (long)a[i] * b[i];
        }
        long end = System.nanoTime();

        System.out.printf("int dot product time: %.6f s (sum: %d)\n", (end - start) / 1e9, sum);
    }

    static void float_dot_product(int size) {
        float[] a = new float[size];
        float[] b = new float[size];

        pcg_state = System.currentTimeMillis();

        for (int i = 0; i < size; i++) {
            a[i] = generate_float();
            b[i] = generate_float();
        }

        long start = System.nanoTime();
        double sum = 0.0;
        for (int i = 0; i < size; i++) {
            sum += (double)a[i] * b[i];
        }
        long end = System.nanoTime();

        System.out.printf("float dot product time: %.6f s (sum: %f)\n", (end - start) / 1e9, sum);
    }

    static void double_dot_product(int size) {
        double[] a = new double[size];
        double[] b = new double[size];

        pcg_state = System.currentTimeMillis();

        for (int i = 0; i < size; i++) {
            a[i] = generate_double();
            b[i] = generate_double();
        }

        long start = System.nanoTime();
        double sum = 0.0;
        for (int i = 0; i < size; i++) {
            sum += a[i] * b[i];
        }
        long end = System.nanoTime();

        System.out.printf("double dot product time: %.6f s (sum: %f)\n", (end - start) / 1e9, sum);
    }

    static void signed_char_dot_product(int size) {
        byte[] a = new byte[size];  // Java byte is signed char in C
        byte[] b = new byte[size];

        pcg_state = System.currentTimeMillis();

        for (int i = 0; i < size; i++) {
            a[i] = generate_signed_char();
            b[i] = generate_signed_char();
        }

        long start = System.nanoTime();
        long sum = 0;
        for (int i = 0; i < size; i++) {
            sum += (long)a[i] * b[i];
        }
        long end = System.nanoTime();

        System.out.printf("signed char dot product time: %.6f s (sum: %d)\n", (end - start) / 1e9, sum);
    }

    static void short_dot_product(int size) {
        short[] a = new short[size];
        short[] b = new short[size];

        pcg_state = System.currentTimeMillis();

        for (int i = 0; i < size; i++) {
            a[i] = generate_short();
            b[i] = generate_short();
        }

        long start = System.nanoTime();
        long sum = 0;
        for (int i = 0; i < size; i++) {
            sum += (long)a[i] * b[i];
        }
        long end = System.nanoTime();

        System.out.printf("short dot product time: %.6f s (sum: %d)\n", (end - start) / 1e9, sum);
    }
}

File: project2/dotproduct.c
#define _POSIX_C_SOURCE 199309L
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdint.h> 

typedef struct timespec TimeSpec;

double get_time_diff(TimeSpec start, TimeSpec end) {
    return (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec)/1e9;
}

int check_and_parse_size(const char *str, int *val) {
    char *endptr;
    if(!str||!val){
        return 0;
    }
    long temp = strtol(str, &endptr, 10);
    
    if (endptr == str || (*endptr != '\n' && *endptr != '\0')) {
        return 0;
    }
    if (temp < 0) {
        return 0;
    }
    
    *val = (int)temp;
    return 1;
}

void int_dot_product(int size);
void float_dot_product(int size);
void double_dot_product(int size);
void signed_char_dot_product(int size);
void short_dot_product(int size);

// 定义 PCG 的状态结构体
typedef struct { 
    uint64_t state;  
    uint64_t inc; 
} pcg32_random_t;

// 初始化一个全局实例
pcg32_random_t rng = {0x853c49e6748fea9bULL, 0xda3e39cb94b95bdbULL};

// 生成 32 位无符号随机数
uint32_t pcg32_random_r() {
    uint64_t oldstate = rng.state;
    rng.state = oldstate * 6364136223846793005ULL + (rng.inc | 1);
    uint32_t xorshifted = ((oldstate >> 18u) ^ oldstate) >> 27u;
    uint32_t rot = oldstate >> 59u;
    return (xorshifted >> rot) | (xorshifted << ((-rot) & 31));
}

// 封装成各类基本数据类型的随机生成函数
int generate_int() {
    return (int)(pcg32_random_r() % 21) - 10;
}

float generate_float() {
    return (float)generate_int() + (float)(pcg32_random_r() % 1000) / 1000.0f;
}

double generate_double() {
    return (double)generate_int() + (double)(pcg32_random_r() % 1000000) / 1000000.0;
}

signed char generate_signed_char() {
    return (signed char)generate_int();
}

short generate_short() {
    return (short)generate_int();
}

int main() {
    char line[100];
    int type;

    // 预定义的需要测试的所有 size 数组
    int sizes[] = {10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);

    while (1) {
        printf("Please select the data type for calculation:\n");
        printf("0. Exit program\n");
        printf("1. int\n");
        printf("2. float\n");
        printf("3. double\n");
        printf("4. signed char\n");
        printf("5. short\n");
        printf("6. all\n");
        printf("> ");

        if (!fgets(line, sizeof(line), stdin)) {
            printf("An error occurred.\n");
            continue;
        }

        if (check_and_parse_size(line, &type)) {
            if (type == 0) {
                printf("Exiting program...\n");
                break;
            } else if (type >= 1 && type <= 6) {
                printf("\n[ Type %d selected, running all sizes ]\n", type);
                for (int i = 0; i < num_sizes; i++) {
                    int size = sizes[i];
                    printf("\n--- Size: %d ---\n", size);
                    switch (type) {
                        case 1: int_dot_product(size); break;
                        case 2: float_dot_product(size); break;
                        case 3: double_dot_product(size); break;
                        case 4: signed_char_dot_product(size); break;
                        case 5: short_dot_product(size); break;
                        case 6:
                            int_dot_product(size);
                            float_dot_product(size);
                            double_dot_product(size);
                            signed_char_dot_product(size);
                            short_dot_product(size);
                            break;
                        default: break;
                    }
                }
                printf("\n[ All sizes computed for type %d. Waiting for next instruction. ]\n", type);
                continue;
            }
        }
        printf("Invalid input. Please enter an integer from 0 to 6.\n");
    }

    return 0;
}
void int_dot_product(int size) {
    int *a = (int *)malloc(size * sizeof(int));
    int *b = (int *)malloc(size * sizeof(int));
    if (!a || !b) { if(a) free(a); if(b) free(b); return; }
    
    rng.state = (uint64_t)time(NULL); 
    for (int i = 0; i < size; i++) {
        a[i] = generate_int();
        b[i] = generate_int();
    }

    long long sum = 0;
    TimeSpec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    for (int i = 0; i < size; i++) {
        sum += (long long)a[i] * b[i];
    }
    clock_gettime(CLOCK_MONOTONIC, &end);
    
    printf("int dot product time: %.6f s (sum: %lld)\n", get_time_diff(start, end), sum);
    free(a); free(b);
}

void float_dot_product(int size) {
    float *a = (float *)malloc(size * sizeof(float));
    float *b = (float *)malloc(size * sizeof(float));
    if (!a || !b) { if(a) free(a); if(b) free(b); return; }
    
    rng.state = (uint64_t)time(NULL);
    for (int i = 0; i < size; i++) {
        a[i] = generate_float();
        b[i] = generate_float();
    }

    float sum = 0;
    TimeSpec start, end;

    clock_gettime(CLOCK_MONOTONIC, &start);
    for (int i = 0; i < size; i++) {
        sum += a[i] * b[i];
    }
    clock_gettime(CLOCK_MONOTONIC, &end);
    
    printf("float dot product time: %.6f s (sum: %f)\n", get_time_diff(start, end), sum);
    free(a); free(b);
}

void double_dot_product(int size) {
    double *a = (double *)malloc(size * sizeof(double));
    double *b = (double *)malloc(size * sizeof(double));
    if (!a || !b) { if(a) free(a); if(b) free(b); return; }
    
    rng.state = (uint64_t)time(NULL);

    for (int i = 0; i < size; i++) {
        a[i] = generate_double();
        b[i] = generate_double();
    }

    double sum = 0;
    TimeSpec start, end;

    clock_gettime(CLOCK_MONOTONIC, &start);
    for (int i = 0; i < size; i++) {
        sum += a[i] * b[i];
    }
    clock_gettime(CLOCK_MONOTONIC, &end);
    
    printf("double dot product time: %.6f s (sum: %f)\n", get_time_diff(start, end), sum);
    free(a); free(b);
}

void signed_char_dot_product(int size) {
    signed char *a = (signed char *)malloc(size * sizeof(signed char));
    signed char *b = (signed char *)malloc(size * sizeof(signed char));
    if (!a || !b) { if(a) free(a); if(b) free(b); return; }
    
    rng.state = (uint64_t)time(NULL);

    for (int i = 0; i < size; i++) {
        a[i] = generate_signed_char();
        b[i] = generate_signed_char();
    }

    long long sum = 0;
    TimeSpec start, end;

    clock_gettime(CLOCK_MONOTONIC, &start);
    for (int i = 0; i < size; i++) {
        sum += (long long)(a[i] * b[i]);
    }
    clock_gettime(CLOCK_MONOTONIC, &end);
    
    printf("signed char dot product time: %.6f s (sum: %lld)\n", get_time_diff(start, end), sum);
    free(a); free(b);
}

void short_dot_product(int size) {
    short *a = (short *)malloc(size * sizeof(short));
    short *b = (short *)malloc(size * sizeof(short));
    if (!a || !b) { if(a) free(a); if(b) free(b); return; }
    
    rng.state = (uint64_t)time(NULL);

    for (int i = 0; i < size; i++) {
        a[i] = generate_short();
        b[i] = generate_short();
    }

    long long sum = 0;
    TimeSpec start, end;

    clock_gettime(CLOCK_MONOTONIC, &start);
    for (int i = 0; i < size; i++) {
        sum += (long long)(a[i] * b[i]);
    }
    clock_gettime(CLOCK_MONOTONIC, &end);
    
    printf("short dot product time: %.6f s (sum: %lld)\n", get_time_diff(start, end), sum);
    free(a); free(b);
}